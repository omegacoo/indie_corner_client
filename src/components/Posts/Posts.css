.Posts {
    border: 1px solid black;
    height: calc(90vh - 4px); /* Remove after initial set up */
}

.Forum-name {
    text-align: center;
}

.Posts-list {
    overflow: auto;
    height: 80%;
    width: 75%;
}

.post,
.new-post {
    border: 1px solid black;
    width: 92.5%;
    list-style: none;
    margin: 2vh;
}

.post h3,
.post span,
.post p,
.new-post-content {
    margin: 2vh;
}

.delete-post {
    border: 1px solid black;
    background-color: white;
    float: right;
    margin: .5vh;
}

.new-post label {
    display: block;
    margin-left: 2vw;
    margin-top: 1vh;
}

.new-post-button {
    border: 1px solid black;
    background-color: white;
    margin: .5vh;
    margin-left: 77%;
}

.back-to-forums-button {
    border: 1px solid black;
    text-decoration: none;
    color: black;
    display: block;
    margin-left: auto;
    margin-right: auto;
    width: fit-content;
    padding: .25vh;
}