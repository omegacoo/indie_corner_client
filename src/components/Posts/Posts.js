import React from 'react';
import { Link } from 'react-router-dom';
import config from '../../config';

import StoreContext from '../../StoreContext';

import './Posts.css';

export default class Posts extends React.Component {
    static contextType = StoreContext;

    state = {
        posts: [],
        content: ''
    };

    componentDidMount(){
        this.fetchPosts();
    }

    fetchPosts = () => {
        fetch(config.API_ENDPOINT + `/api/posts/${this.context.currentForum}`)
            .then(res => {
                if(!res.ok){
                    throw new Error(res.statusText);
                }
                return res.json();
            })
            .then(resJson => {
                this.setState({
                    posts: resJson
                })
            })
            .catch(err => {
                console.log(err);
            })
    }

    getCurrentTime(){
        let today = new Date();
        let date = today.getFullYear() + '-' + (today.getMonth() + 1) + '-' + today.getDate();
        let time = today.getHours() + ':' + today.getMinutes() + ':' + today.getSeconds();
        let dateTime = date + ' ' + time;
    
        return dateTime;
    }

    handleRemovePost = id => {
        const cookie = document.cookie.replace(/(?:(?:^|.*;\s*)token\s*=\s*([^;]*).*$)|^.*$/, "$1");
        const myHeaders = new Headers();
        myHeaders.append('Cookies', cookie);

        fetch(config.API_ENDPOINT + `/api/posts/remove_post/${id}`, { method: 'DELETE', headers: myHeaders })
            .then(res => {
                if(!res.ok){
                    throw new Error(res.statusText);
                }
            })
            .then(() => {
                this.fetchPosts();
            })
            .catch(err => {
                console.log(err);
                alert('You must be logged in!');
            })        
    }

    handleNewPostSubmit = e => {
        e.preventDefault();

        const cookie = document.cookie.replace(/(?:(?:^|.*;\s*)token\s*=\s*([^;]*).*$)|^.*$/, "$1");
        const myHeaders = new Headers();
        myHeaders.append('Content-Type', "application/json");
        myHeaders.append('Cookies', cookie);
        const newPost = JSON.stringify({
            user_id: this.context.userId,
            forum_id: this.context.currentForum,
            content: this.state.content,
            time_submitted: this.getCurrentTime()
        });

        fetch(config.API_ENDPOINT + `/api/posts/${this.context.currentForum}`, { method: 'POST', body: newPost, headers: myHeaders })
            .then(res => {
                if(!res.ok){
                    throw new Error(res.statusText);
                }
                this.fetchPosts();
            })
            .catch(err => {
                console.log(err);
                alert('You must be logged in!');
            })
        
        this.setState({
            content: ''
        })
    }

    handleContentChange = e => {
        this.setState({
            content: e.target.value
        });
    }

    getForumName = () => {
        let forum = this.context.forums.find(forum => forum.id === this.context.currentForum);

        return forum.title;
    }

    renderPosts = () => {
        let forumPosts = this.state.posts.filter(post => post.forum_id === this.context.currentForum);

        return forumPosts.map(post => 
            <li 
                key={post.id}
                className='post'
            >
                <button
                    className='delete-post'
                    onClick={() => this.handleRemovePost(post.id)}
                >
                    delete
                </button>
                <h3>{post.user_name}</h3>
                <p>{post.content}</p>
                <span>{post.time_submitted}</span>
            </li>
        )
    }

    render(){
        return(
            <div className='Posts'>
                <h2 className='Forum-name'>{this.getForumName()}</h2>
                <ul className='Posts-list'>
                    { this.renderPosts() }
                    <li className='new-post'>
                        <form 
                            className='new-post-form'
                            id='new-post-form'
                            onSubmit={this.handleNewPostSubmit}
                        >                            
                            <label htmlFor='new-post-content'>New Post</label>
                            <input 
                                className='new-post-content' 
                                onChange={this.handleContentChange}
                                value={this.state.content}
                            />
                            
                            <button 
                                className='new-post-button' 
                                type='submit'
                            >
                                submit
                            </button>
                        </form>
                    </li>
                </ul>
                <Link 
                    className='back-to-forums-button'
                    to={'/forums'}
                >
                    Forums
                </Link>
            </div>
        );
    };
};