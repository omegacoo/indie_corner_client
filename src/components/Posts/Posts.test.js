import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';
import Posts from './Posts';

describe('Posts component', () => {
    let context = {
        currentForum: 1,
        currentUser: 'anonymous'
    };

    it('renders without crashing', () => {
        const div = document.createElement('div');

        ReactDOM.render(
            <BrowserRouter>
                <Posts context={context}/>
            </BrowserRouter>, 
            div
        );
        ReactDOM.unmountComponentAtNode(div);
    });

    it('renders UI correctly', () => {
        expect(
            renderer.create(
                <BrowserRouter>
                    <Posts context={context}/>
                </BrowserRouter>
            ).toJSON()
        ).toMatchSnapshot();
    });
});