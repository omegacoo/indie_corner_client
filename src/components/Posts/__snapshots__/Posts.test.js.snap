// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Posts component renders UI correctly 1`] = `
<div
  className="Posts"
>
  <h2
    className="Forum-name"
  >
    One
  </h2>
  <ul
    className="Posts-list"
  >
    <li
      className="post"
    >
      <button
        className="delete-post"
        onClick={[Function]}
      >
        X
      </button>
      <h3>
        Kris
      </h3>
      <span>
        2020-5-28 11:13:40
      </span>
      <p>
        This is the first post Kris made
      </p>
    </li>
    <li
      className="post"
    >
      <button
        className="delete-post"
        onClick={[Function]}
      >
        X
      </button>
      <h3>
        Riley
      </h3>
      <span>
        2020-5-28 11:13:40
      </span>
      <p>
        Riley cant read, her dad must have made this one
      </p>
    </li>
    <li
      className="post"
    >
      <button
        className="delete-post"
        onClick={[Function]}
      >
        X
      </button>
      <h3>
        Coco
      </h3>
      <span>
        2020-5-28 11:13:40
      </span>
      <p>
        Coco is a dog, who made this for her?
      </p>
    </li>
    <li
      className="post"
    >
      <button
        className="delete-post"
        onClick={[Function]}
      >
        X
      </button>
      <h3>
        Bill
      </h3>
      <span>
        2020-5-28 11:13:40
      </span>
      <p>
        Mr. Gates checking in
      </p>
    </li>
    <li
      className="new-post"
    >
      <form
        className="new-post-form"
        onSubmit={[Function]}
      >
        <label
          htmlFor="new-post-content"
        >
          Post
        </label>
        <input
          className="new-post-content"
          onChange={[Function]}
          value=""
        />
        <button
          className="new-post-button"
          type="submit"
        >
          submit
        </button>
      </form>
    </li>
  </ul>
  <a
    className="back-to-forums-button"
    href="/forums"
    onClick={[Function]}
  >
    Forums
  </a>
</div>
`;
